hi-ml/src/health_ml/networks/layers/attention_layers.py

```diff
@@ -89,3 +86,161 @@ def forward(self, features: Tensor) -> Tuple[Tensor, Tensor]:
+            x = x + self._ff_block(self.norm2(x))
+        else:
+            sa_block_out, a = self._sa_block(x, src_mask, src_key_padding_mask)
+            x = self.norm1(x + sa_block_out)
+            x = self.norm2(x + self._ff_block(x))
+
+        return x, a
+
+    # self-attention block
+    def _sa_block(self, x: Tensor,
+                  attn_mask: Optional[Tensor], key_padding_mask: Optional[Tensor]) -> Tuple[torch.Tensor, torch.Tensor]:
+        x, a = self.self_attn(x, x, x,
+                              attn_mask=attn_mask,
+                              key_padding_mask=key_padding_mask,
+                              need_weights=True)  # Just because of this flag I had to copy all of the code...
+        x = x[0]
+        return self.dropout1(x), a
+
+    # feed forward block
+    def _ff_block(self, x: Tensor) -> Tensor:
+        x = self.linear2(self.dropout(self.activation(self.linear1(x))))
+        return self.dropout2(x)
+
+
+class TransformerPooling(Module):
+    """Create a Transformer encoder module consisting of multiple Transformer encoder layers.
+
+    We use a additional classification token (cls token) for pooling like seen in ViT/Bert. First, the cls token is
+    appended to the list of tiles encodings. Second, we perform self-attention between all tile encodings and the cls
+    token. Last, we extract the cls token and use it for classification.
+
+    Args:
+        num_layers: Number of Transformer encoder layers.
+        num_heads: Number of attention heads per layer.
+        dim_representation: Dimension of input encoding.
+    """
+    def __init__(self, num_layers: int, num_heads: int, dim_representation: int) -> None:
+        super(TransformerPooling, self).__init__()
+        self.num_layers = num_layers
+        self.num_heads = num_heads
+        self.dim_representation = dim_representation
+
+        self.cls_token = nn.Parameter(torch.zeros(dim_representation))
+
+        self.transformer_encoder_layers = []
+        for _ in range(self.num_layers):
+            self.transformer_encoder_layers.append(
+                CustomTransformerEncoderLayer(self.dim_representation,
+                                              self.num_heads,
+                                              dim_feedforward=self.dim_representation,
+                                              dropout=0.1,
+                                              activation=F.gelu,
+                                              batch_first=True))
+        self.transformer_encoder_layers = torch.nn.ModuleList(self.transformer_encoder_layers)  # type: ignore
+
+    def forward(self, features: Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        cls_token = (torch.ones(1, self.dim_representation, device=features.device) * self.cls_token)
+
+        # Append cls token
+        H = torch.vstack([cls_token, features]).unsqueeze(0)
+
+        for i in range(self.num_layers):
+            H, A = self.transformer_encoder_layers[i](H)
```

Comment:
What are the shapes of `H` and `A`? It would be helpful to mention this in a comment.